[Roach] 미션 5: 보드에 위치 부여 및 점수 계산

tmdgusya opened the pull request:
이번 요구사항은 조금 어려워서 놓친것들이 세세하게 많을 수도 있다고 생각이듭니다.

최대한 요구사항에 준수하기 위해 노력했으며, 메소드 중복 또는 분리를 위해서 리팩토링을 진행하였습니다.
다만 인터페이스를 적용할 부분을 찾아보고 코드가 복잡해질지, 코드가 더 깔끔해 지는지를 계산해보라고 하셨는데
제 생각에는 각 class 내 Field 의 의존성이 높아서 인터페이스로 분리한들 별로 깔끔해지지 않겠다라는 생각이 들었습니다.
지금의 코드들과 별로 다르지 않을꺼라고 판단되었습니다. (제 기준에서는요 ! 리뷰어님들의 의견도 듣고 공부해보고 싶습니다. )
저번에 상코의 리뷰를 듣고 최대한 분리를 해보려고 했는데 잘 분리된것인지는 모르겠네요.

리뷰어님들의 리뷰 부탁드립니다!

tmdgusya committed:
feat : 기물의 색, 기물의 종류에 따른 enum 구현
Color.java
- Color : 요구사항에 맞게 NOCOLOR 라는 새로운 속성을 추가하였음.

Type.java
- 요구사항에 맞게 각 기물들이 가지고 있는 representation 을 속성값으로 가지도록 설정하였음.

Piece.java
- Type representation 값을 이용해야 하므로 기존 representation 상수 열거들제가하고 Enum Type 으로 변경

tmdgusya committed:
refactor : 코드 일부 변경

Pieces.java
- validateSize() : 해당 메소드 자체에서 Exception 을 return 하도록 변경하였음. add 메소드에서 에러를 리턴할 이유가 없다고 생각되어서.

Board.java
- 각종 속성들이 Enum 에 추가되면서 해당 속성에 따른 처리를 if 문으로 계속 처리해주기에는 복잡도가 높아진다고 판단되어, 메소드 분리를 실천함.

tmdgusya committed:
feat : createBlank() 메소드 추가

요구사항으로 인해 Blank Piece 가 필요로 해지면서 해당 Piece 도 팩토리 메소드를 통한 공급을 가능하도록 만들었음.

* 해당 메소드 단위 테스트도 PieceTest.java 에 추가하였음.

tmdgusya committed:
refactor : 팩토리 메소드 리팩토링

- 팩토리 메소드에 다수의 중복코드인 new Piece 를 색깔별로 구분한뒤 각 Type 의 Piece 로서 만들어주는 createWhite / createBlack 을 추가하여 중복 코드를 제거하였음.

tmdgusya committed:
feat : 기물과 색에 해당하는 기물의 개수를 반환 및 주어진 위치의 기물을 조회

1. Type.java
- NO_PIECE 요구사항과 같이 x -> . 으로 변경하였음

2. RANK.java
- 기존에 사용하던 Pieces 를 RANK 로 대체하였음. 요구사항으로 보았을때 RANK 에 알맞다고 생각이들기 때문에.
- countOfBlackPieces / countOfWhitePieces : 각 Piece 색깔별로 개수를 구할 수 있도록 하였음.
- countOfPiece : Color 와 Type 으로 개수를 구할수 있도록 하였음. 원래는 각 Type 별로 메소드 분리로 했으나, 굳이 분리 안하고 현재와 같은 방식이 복잡도도 낮추고 더 깔끔하다고 판단했음.

3. Position.java
- 각 랭크와 인덱스에 필요한 처리 계산과정을 해당 클래스에서 행하도록 함.

4. Piece.java

- hashCode() 와 equals 의 재정의 : VO 객체라 동등성을 비교하기 위해서 해당 메소드를 재정의 해주었음.
- private 메소드 밑으로 이동

5. Board.java

- countOfPiece : 보드 위에서 각 피스의 Count 를 구할 수 있도록 함.
- findPiece : 해당 position 에서 거친 연산결과를 기반으로 piece 를 리턴해주는 구조이다.
- getBoardStatusToString : board 자체를 커스텀 자료구조로 관리하게 되면서 편리하게 바뀜.

tmdgusya committed:
test : 기물과 색에 해당하는 기물의 개수를 반환 및 주어진 위치의 기물을 조회 테스트 생성

BoardTest.java

- Board 에서 각 Type 별로 Piece 의 Count 를 정확히 계산하는지를 확인하는 테스트를 만들었음
- checkFindPieceMethod : 원하는 piece 를 정확히 찾아오는지 테스트

tmdgusya committed:
feat : 임의의 기물을 체스판 위에 추가

Board.java
- initializeEmpty() : 체스의 모든 칸을 공백으로 체우는 메소드 추가
- move() : board 에서 원하는 위치에 말을 놓는 메소드 추가

Rank.java
1. RANK -> Rank 로 클래스 네이밍 변경
2. putPiece() : 자료구조 상에서 원하는 위치에 말을 놓을 수 있도록 메소드 추가.

tmdgusya committed:
test : 임의의 기물을 체스판 위에 추가 TEST 파일 추가

BoardTest.java
- checkCorrectPosition() : 원하는 위치에 말이 정상적으로 놓여지는지 테스트하기 위해 추가함.

tmdgusya committed:
feat : 체스 프로그램 점수 계산하기

1. Type.java
새로운 속성인 각 체스말마다 가질수 있는 point 추가

2. Rank.java
- 원래 랭크와 인덱스를 바꿔서 하다가 그냥 랭크별로 하는걸로 정했음. 그게 훨씬 편하고 복잡한 처리과정이 필요하지 않음
- 각 색깔별 Piece 에 대한 점수를 구할 수 있도록함
- getPointsOfType :
  - 람다식을 통해 간결한 방식으로 포인트를 구할 수 있도록 하였음.

3. Piece.java
 - 출력을 위해서 toString 재정의

4. Board.java
- 각 팀에 대한 점수를 구할 수 있는 메소드를 추가하였음.
  - blackTeamPoint / whiteTeamPoint
- 단 요구사항에서 같은 랭크의 폰의 경우 1/2 의 삭감이 이뤄져야 하므로 해당 로직을 samePlacedPawns 에서 진행하였음.
- getBoardStatusToString : 기존 (Index, Rank) 방식에서 (Rank, Index) 방식으로 변환하면서 출력문도 변경하였음.

tmdgusya committed:
feat : 기물의 점수가 높은 순으로 정렬

1. Piece.java
 - 기물의 정렬을 위해서 point field 하나와 compareTo 를 추가하였음.

tmdgusya committed:
test : 점수 테스트 추가 및 정렬 테스트 추가

1. BoardTest.java
- checkPointOfTeam() : 해당 메소드를 통해서 점수가 잘 계산되는지 확인

2. RankTest.java

- sortToPieces() :  Piece 들이 Point 별로 정렬되는지 확인

tmdgusya committed:
refactor : 중복코드제거

기존의 색깔별로 Piece Count 를 세던것의 메소드들에서 중복코드가 일어나고 있기에 Rank.java 의 countOfTeamPieces() 를 통해서 중복을 줄일 수 있도록 하였음.

wheejuni commented on src/main/java/net/honux/chess/attribute/Color.java:
```diff
@@ -1,5 +1,5 @@
 
 public enum Color {
-    WHITE, BLACK
+    WHITE, BLACK, NOCOLOR;
```
`NOCOLOR` 라는 접근법 좋은 것 같습니다. 
그런데 `enum`은 한 줄에 하나씩 선언해주세요.

```suggestion
        WHITE,
        BLACK,
        NOCOLOR;

tmdgusya replied:
감사합니다. 혹시 가독성때문인가요?
아니면 관례같은 것인가요?

tmdgusya replied:
오우 해보니 훨씬더 깔끔해보이기는 하네요 감사해요 👍

wheejuni commented on src/main/java/net/honux/chess/attribute/Type.java:
```diff
@@ -0,0 +1,25 @@
+
+public enum Type {
+    PAWN('p', 1), ROOK('r', 5), KNIGHT('n', 2.5),
+    BISHOP('b', 3), QUEEN('q', 9), KING('k', 0), NO_PIECE('.', 0);
```
여기도 같은 의견 드립니다.

wheejuni commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,106 +1,121 @@
-    private Pieces whitePawns;
-    private Pieces blackPieces;
-    private Pieces blackPawns;
+    private ArrayList<Rank> board;
```
`ArrayList` 를 변수 타입으로 받은 코드들에 달린 다른 피드백을 참고해보시면 좋겠어요.

tmdgusya replied:
네넵 오늘 한번 읽어보고 어떤 부분에 대한 내용들이 나왔는지 정리해보고 적용시켜보도록 할께요!

tmdgusya replied:
일단 리뷰들을 돌아다니며 정리한 바로는
1. board 의 경우 재할당이 되어야 할 이유가 없으므로 final 로 선언하는 것이 좋다. => final 로 적용한뒤 생성자에서 해당 구현체를 대입해주자.
2. ArrayList<Rank> board; => List<Rank> board; : 이 부분은 제가 List 로 선언했었는데 .. 또 무언가 하다가 바뀌었는데 인지하지 못했군요. 이런 기본적인걸 놓쳐서 Commit 하다니 앞으로는 조금 더 신중하게 봐야겠다는 생각이 듭니다.

wheejuni commented on src/main/java/net/honux/chess/core/Rank.java:
```diff
@@ -0,0 +1,196 @@
+        return countOfWhitePieces();
+    }
+
+    public int countOfBlackPieces() {
+        return (int) pieceList.stream().filter(piece -> Character.isUpperCase(piece.getRepresentation())).count();
+    }
+
+    public int countOfWhitePieces() {
+        return (int) pieceList.stream().filter(piece -> Character.isLowerCase(piece.getRepresentation())).count();
+    }
```
`Piece` 에게 색깔도 물어보면 더 간결한 코드로 개선할 수 있지 않을까요?

tmdgusya replied:
감사해요! :)

아래와 같이 바꿨더니 가독성면이나 간결성에서 훨씬 더 좋은 코드로 변신했네요.
```java
return (int) pieceList.stream().filter(Piece::isBlack).count();
```

wheejuni commented on src/main/java/net/honux/chess/core/Rank.java:
```diff
@@ -0,0 +1,196 @@
+        return (int) pieceList.stream().filter(piece -> Character.isLowerCase(piece.getRepresentation())).count();
+    }
+
+    public int countOfPiece(Color color, Type type) {
+        return (int) pieceList.stream().filter(piece -> piece.getRepresentation() == type.getRepresentation(color)).count();
+    }
```
`piece.isWhite()` 라는 아름다운 메소드가 있는 것 같은데요, 
이걸 이용해서 더 좋은 코드로 개선할 수 있을 것 같은데, 어떻게 생각하세요?

tmdgusya replied:
이 부분에서 piece.isWhite() 를 어떻게 써야할지 잘 모르겠습니다.
제 생각에는 여기서 representation 에 대한 비교가 일어난다고 일단 생각을 하였는데
혹시 약간의 힌트를 주실 수 있으신가요?

wheejuni replied:
지금 우리의 요구사항에서 `Color`는 두 종류이죠?
아 죄송합니다;; 하나 더 있네요 노컬러
어쨌든, `Piece` 에게 색깔을 물어보는 방향의 구현이 이뤄지면 좋겠습니다.

tmdgusya replied:
이런식으로 바꿔봤는데 어떻게 생각하시나요?
```java
 public int countOfPiece(Color color, Type type) { // Original
        return (int) pieceList.stream().filter(piece -> piece.getRepresentation() == type.getRepresentation(color)).count();
    }

    public int countOfWhiteTypePieces(Color color, Type type) { // New
        return (int) pieceList.stream().filter(piece -> piece.isMatchingColor(color)).filter(piece -> piece.isMatchingRepresentation(color, type)).count();
    }
```

tmdgusya replied:
아 아래걸 바꿔보니 이렇게 해도되네요
```java
public int countOfWhiteTypePieces(Color color, Type type) {
     return (int) pieceList.stream().filter(piece -> piece.isMatchingRepresentation(color, type)).count();
}
```

```java
public boolean isMatchingRepresentation(Color color, Type type) {
    return this.representation == type.getRepresentation(color);
}
```

해당 Color 별 Type에 대한 Representation 을 가져오고, 현재 Piece 의 Representation 최대한 Piece 에 색깔과 Color 를 물어보는 방향으로 코드를 바꿔보았습니다. 확실히 객체 지향상으로도 기존의 코드에서 비교문을 쓰기보다는 Brian 말대로 객체 자체에서 piece.isMatchingRepresentation(color, type) 색깔과 Type 에 대한 boolean 을 주는것이 좋은것 같습니다 !

wheejuni commented on src/main/java/net/honux/chess/core/Rank.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    public int getPointsOfType(Color color) {
+        int point = 0;
+        for (Type type : Type.values()) {
+            point += pieceList.stream().filter(piece -> piece.getRepresentation() == type.getRepresentation(color)).count() * type.getPoint();
+        }
+        return point;
```
재밌는 개선포인트가 있을 것 같습니다. 
`color` 베이스로 `pieceList` 를 필터한 다음, `Piece` 를 `Type` 으로 바꾼 후, 다시 `Type` 을 `double` 로 바꿔서 합계를 구하면 좋은 스트림 기반 코드 한 편이 나올 것 같네요. 

_학습 키워드_
> `.filter()`
> `.map()`, `.mapToDouble()`
> `.sum()`

tmdgusya replied:
와우 람다식을 이렇게 바꿔볼 수 있었어요!
```java
public int getPointsOfType(Color color) {
        return (int) pieceList.stream().filter(piece -> piece.isMatchingColor(color)).mapToDouble(Piece::getPoint).sum();
    }
```

tmdgusya replied:
진짜 간결하게 바뀐것 같네요. :)

wheejuni commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,106 +1,121 @@
+    }
+
+    public Optional<Piece> findPiece(String position) {
+        Optional<Piece> piece = Optional.empty();
```
`Optional` 의 잘못된 용법입니다. `Optional.empty()` 는 `return` 문과 함께 사용할때 가장 아름답습니다. 
불필요한 변수로 보이고 적절한 시점에서의 `return` 을 통해 리팩토링할 수 있을 것 같은데 고민해주세요.

tmdgusya replied:
오 이렇게 사용하니 적절하게 리팩토링 된것 같습니다!
이런 세세한 부분은 생각하지 못했네요 감사해요 👍 
```java
 public Optional<Piece> findPiece(String position) {
        try {
            return Optional.ofNullable(board.get(getRank(position)).get(getIndex(position)));
        } catch (IndexOutOfBoundsException e) {
            return Optional.empty();
        }
    }
```

wheejuni commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,106 +1,121 @@
-    public Optional<Piece> findBlackPiece(int index) {
-        Optional<Piece> pawn = Optional.empty();
+    public int countOfPiece(Color color, Type type) {
+        int countOfBlackPawns = 0;
+        for (Rank rank : board) {
+            countOfBlackPawns += rank.countOfPiece(color, type);
+        }
+        return countOfBlackPawns;
+    }
```
[이 글](https://www.baeldung.com/java-stream-filter-count) 을 읽고 리팩토링 아이디어를 얻어보세요. 
더 간단하게 줄일 수 있습니다.

tmdgusya replied:
아래와 같이 단축해보았습니다 !
```java
return (int) board.stream().mapToInt(rank -> rank.countOfPiece(color, type)).sum();
```

wheejuni commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,106 +1,121 @@
-        appendNewLine(sb, getStringPiecesResult(blackPawns));
-        for (int i = 0; i < initialSpaceInterval; i++) {
-            appendNewLine(sb, BLANK);
+        for (int i = 0; i < BOARD_SIZE; i++) {
+            for (int j = 0; j < BOARD_SIZE; j++) {
+                sb.append(board.get(j).get(i).toString());
+            }
+            appendNewLine(sb, "");
```
이중 `for loop` 는 많은 경우에 code smell에 해당합니다.
개선할 수 없을까요?

tmdgusya replied:
next Position 에 대한 값들을 return 해주는 방식으로 진행하면 될것 같은데 조금 더 고민을 해봐야 할 것 같아요.
이중 루프를 지우는 것이 쉬운일이 아니네요

tmdgusya replied:
RANK 라는 Enum 을 도입하고 체스 기물들이 가질수 있는 Position 을 갖도록 하여 처리하였습니다.
이번 리뷰 관련된 내용 다시 commit 올리도록 하겠습니다.

부족한 부분이나 좋지않은 방향으로 개선된 부분 알려주시면 감사하겠습니다 👍

wheejuni reviewed (changes requested):
5단계까지 진행하시느라 고생 많으셨습니다. 💯 
진도를 다소 빠르게 나가고 있는 상황이기때문에, 다음 단계로 급히 넘어가기보단 이번 단계에서 그동안 축적되어 있던 리팩토링 포인트를 모두 해결하고 넘어가는게 좋겠군요. 
이번 리뷰 과정은 조금 험난할 수도 있겠습니다. 

아홉개의 코멘트를 통해 개선 의견을 드렸는데 우선 `Piece` 에 구현되어야 할 메소드가 하나 누락되어있는 것 같고, 여기서부터 깨끗한 설계를 하지 못하게 된 것 같습니다. 
메소드의 namespace와 리턴 타입을 아래와 같이 명시해 드리니 구현해보세요. 

```java
public boolean isMatchingColor(Color color);

tmdgusya committed:
review : 1차 코드리뷰 반영

1. Type.java / Color.java
 - 가독성을 위해서 한줄로 표기
2. Piece.java
 - isMatchingColor : Piece 의 현재색과 매개변수의 현재색이 동일한 색인지 boolean Type 으로 리턴
 - isMatchingRepresentation : Piece 의 현재색과 representation 에 대하 매개변수의 값과 동일한지boolean Type 으로 리턴
3. Position.java
 - 대문자로 들어올시에도 소문자로 바꿔줘야 하므로 getRank 로직 변경
4. Board.java
 - ArrayList -> List 로 변경 : 추상체 타입으로 타입을 선언해야 후에 자율성 부분에서 이점을 얻을 수 있음.
 - board final 로 선언 : 체스가 시작되고 나면 재할당되야할 이유가 없으므로 final 로 선언하였음.
 - getTeamPoints / countOfTeamPieces / countOfPiece : 길었던 메소드 stream 을 이용하여 축약하였음.
 - findPiece() : pawn 에 처음부터 empty 를 대입하는건 좋지 않은 코드 방식이라 생각하여 return 할때 Optional.empty() 를 하도록 함.
 - getBoardStatusToString() : File 이라는 Enum 의 도입으로 For loop 형태의 코드 제거
5. Rank.java
 - 람다식 조건문 Piece 객체가 자의적으로 답할수 있도록 변경!
 - getPointsOfType : isMatchingColor 을 이용하여 간단한 람다식으로 최적화함
6. StringUtils.java / ChessGame.java : EOF 추가!
7. File.java
 - 각 Chess File 들이 가질 수 있는 File 가짓수를 Enum 으로 가질 수 있도록 해서, 현재는 Board 에 출력으로 쓰이고 있음.

wheejuni commented on src/main/java/net/honux/chess/entity/pieces/Piece.java:
```diff
@@ -85,12 +93,12 @@ public boolean isWhite() {
 
-    private static Piece createWhite(Type type) {
-        return new Piece(Color.WHITE, type.getRepresentation(Color.WHITE), type.getPoint());
+    public boolean isMatchingColor(Color color) {
+        return this.color == color;
```
👍

tmdgusya replied:
덕분에 2중 루프를 없애는 방향에 대한 생각을 할 수 있었습니다.
그리고 조금 더 아름다운 코드들에 대한 방향성에 대해 생각을 많이 하게된것 같습니다.
덕분에 많이 발전한것 같습니다 감사합니다 👍

wheejuni reviewed (approved):
수고하셨습니다. 👍
